### ISSUE TITLE
HTTPS certificate

### ISSUE THREAD

**Issue by qdrn** [massalabs/thyra#154]
See how we can improve the procedure for certificate generation. During our quick brainstorm we saw two possibilities:
- Use a an entity to validate redirecting .massa site to localhost. But since the .massa domain does not exist it's probably a dead end.
- Generate and add certificate during installation. We need to see if the procedure differ from one OS to the other or for different browsers.

cc @gregLibert

**Comment by qdrn** [comment]
Related: https://web.dev/how-to-use-local-https/
Not very encouraging tbh.

**Comment by SlnPons** [comment]
Adding some details and context as we'll address it soon. 
Please technical details before starting working on it :)

**Context**
This is due to the https certificate that is not natively integrated to Thyra. This error will quickly become problematic as soon as we'll add wallet / crypto app or transactions in Thyra.
It would be good to start identifying how we can handle this nicely.

**Needs**
Install https certificate on Thyra during the installation process

**User journey**
End user installs Thyra on their device
End user opens their default browser
End user types my.massa/thyra/wallet and is redirected to https://my.massa/thyra/wallet/index.html directly without seeing the screenshot above.

**How to**

Identify how we can install the https certificate during the installation process
Make sure we have a solution for all OS identified in the DoD as well as all browsers (if not, list the identified blockers for each)
Try to implement at best; or to explain the process to do it at least.

**Technical details**
TO BE ADDED

**Comment by 0xMazout** [comment]
Collectings data : 
https://www.smashingmagazine.com/2017/06/guide-switching-http-https/
https://www.section.io/engineering-education/how-to-get-ssl-https-for-localhost/



![Image](https://user-images.githubusercontent.com/10938862/200319613-76b793c5-d369-4fe5-acd3-dc51361827f4.png)

**Comment by 0xMazout** [comment]
In the Section.IO tutorial, The windows way to have https on localhost is well explained, we have to follow this process :

- Generate a CA certificate (Will create a CA key [ca.key] and a CA certificate [ca.pem]) 
- Generate a SSL certificate (*.ext file) 
- Generate a CSR (Certificate Signing Request [*.csr])
- Importing CA certificate to the default selected browser ( This is to have the validation from the browser we are a good CA )

On each installation of Thyra we have to do all of this to permit https. 
The last point with the browser is to avoid the warning page 

Limitations : 
- We have to make a different process for all of the OS ( Linux, Darwin, Windows)
- We have to embeded OpenSSL into Thyra

**Comment by 0xMazout** [comment]
Or we can use mkcert : https://github.com/FiloSottile/mkcert

And try to import the CA certificate into the browser , this should delete the warning page on each browsers

**Comment by 0xMazout** [comment]
After discussion with Greg, The main subject is to clarify if yes or no we can change browser settings via cmd line

**Comment by 0xMazout** [comment]
Firefox Certificates : https://github.com/mozilla/policy-templates/blob/master/README.md#certificates

/!\ There is a bug on Linux Version /!\ https://bugzilla.mozilla.org/show_bug.cgi?id=1600509

Compatibility: Firefox 60, Firefox ESR 60 (macOS support in Firefox 63, Firefox ESR 68)

https://support.mozilla.org/fr/kb/setting-certificate-authorities-firefox



![Image](https://user-images.githubusercontent.com/10938862/200343146-54693bd9-d545-4aa2-a9c7-90c19196d8e0.png)

**Comment by 0xMazout** [comment]
Updating Chrome https://chromium.googlesource.com/chromium/src/+/refs/heads/lkgr/docs/linux/cert_management.md

**Comment by 0xMazout** [comment]
How to do it with open SSL
https://stackoverflow.com/a/60516812

**Comment by peterjah** [comment]
@qdrn @0xMazout @SlnPons 
Hello everyone!
I added the certificate creation with mkcert when installing thyra (see PR: https://github.com/massalabs/thyra/pull/288 )
The cert is only valid for the "my.massa" TLD.
As explained in this stackoverflow thread, it is not possible to create a certificate for "*.massa" :

![image](https://user-images.githubusercontent.com/34547263/203357623-249d42f9-3b14-4188-bf5e-48d28bf0b8ba.png)

The solution would be to add a subdomaine TLD for every websites on the blockchain `flappy.massa` would become `flappy.my.massa` for example and then we create a certificat for *.my.massa .. it should work

**Comment by qdrn** [comment]
Unfortunate but better than nothing.

**Comment by gregLibert** [comment]
> @qdrn @0xMazout @SlnPons Hello everyone! I added the certificate creation with mkcert when installing thyra (see PR: #288 ) The cert is only valid for the "my.massa" TLD. As explained in this stackoverflow thread, it is not possible to create a certificate for "*.massa" :
> 
> ![image](https://user-images.githubusercontent.com/34547263/203357623-249d42f9-3b14-4188-bf5e-48d28bf0b8ba.png)
> 
> The solution would be to add a subdomaine TLD for every websites on the blockchain `flappy.massa` would become `flappy.my.massa` for example and then we create a certificat for *.my.massa .. it should work

I was thinking about using go capabilities to generate certificate. But If you prefer mkcert, it's always painful to install a new mandatory dependency, but it's doable.

Regarding the subject of the certificate, I was thinking about:
- using SNI capability to dynamically generate the server certificate using an AC.
- only proposing the certificate for few SAN (the spec doesn't defined any limit, I guess the browser will be the limit) which can be a commercial vector around thyra.

@peterjah wdyt?

**Comment by peterjah** [comment]
The interest of using mkcert is that mkcert itself is a Certificate Authority which is accepted by browsers.
its not so painful to use. just DL mkcert, generate certificate, remove mkcert... the job is already done in the PR i mentionned and its working cross OS.
About SNI (several certificate for 1 IP) it could work, but i guess we will have to do it with mkcert anyway
about SAN (several domain in a single certificate) this could work also but its very "scotch and ficelle" . Do you really want to monetize the SSL capabilities of a deployed website? :/

**Comment by peterjah** [comment]
To be more precise about SAN option: the probleme is we will have to maintain somewhere a list of the certified domains which will be painfull. and also  SAN certificate is limited to 250 granted domains by design...

**Comment by SlnPons** [comment]
Thanks guys for sharing ideas and discussing 💪
As discussed together, we'll test the SNI method and see if it could work. Which means that the scope of this ticket is reduced to: 
* Certified Thyra pages in my.massa

The rest will be tested using this task: https://github.com/massalabs/thyra/issues/307

**Comment by gregLibert** [comment]
> The benefits of using mkcert is that mkcert itself is a Certificate Authority which is accepted by browsers. Which is not the case if you generate the certificates on your own. its not so painful to use. just DL mkcert, generate certificate, remove mkcert... the job is already done in the PR i mentionned and its working cross OS.

From what I understand about mkcert, my source is [here](https://github.com/FiloSottile/mkcert), it is not a browser trusted certificate authority, but a certification tool suite that generates a new certificate authority and adds it to certain root stores such as Java, OS or certain browsers.
It does not take anything away from the mkcert utility, but this approach has its limitations, see [the documentation](https://github.com/FiloSottile/mkcert#supported-root-stores).

> its not so painful to use. just DL mkcert, generate certificate, remove mkcert... the job is already done in the PR i mentionned and its working cross OS.

With the Go programming language, all the tools needed to create certificates are on the shelf. So, the only added value of mkcert is to add the generated certificate to the known root stores.
Painful was not necessarily the right word. When we increase the number of tools we use, we also increase the overall complexity of the project.
But as a developer, it's your job to make these difficult choices :-).

> Do you really want to monetize the SSL capabilities of a deployed website?

No, security is not to be monetize. But VIP user journey could be. For instance we could reserve the subdomain *.partner.massa to VIP partners only and manage this subdomain through SAN.
